File: calculatingdartscores.c
#include<stdio.h>

struct combination{
    int value;
    int a,b,c;
};

void print_combination(int x, int p[]);

int main(){
    int n;
    scanf("%d",&n);
    
    if (n==0){
        printf("impossible");
        return 0;
    }
    
    int p[61];
    for (int i=0; i<60; i++){
        if (i<20) p[i]=(20-i)*3;
        else if (i>=20 && i<40) p[i]=p[i-20]/3*2;
        else p[i]=p[i-20]/2;
    }
    p[60]=0;
    
    struct combination l[61*61*61];
    int l_index=0;
    for (int i=0; i<61; i++)
        for (int j=0; j<61; j++)
            for (int k=0; k<61; k++){
                l[l_index].value=p[i]+p[j]+p[k];
                l[l_index].a=i;
                l[l_index].b=j;
                l[l_index++].c=k;
            }
    int i;
    for (i=0; i<sizeof(l)/sizeof(l[0])-1; i++)
        if (l[i].value == n) break;
    
    if (i==sizeof(l)/sizeof(l[0])-1){
        printf("impossible");
        return 0;
    }
    
    
    int a=l[i].a, b=l[i].b, c=l[i].c;

    print_combination(a,p);
    print_combination(b,p);
    print_combination(c,p);
    
    
    return 0;
}
void print_combination(int x, int p[]){
    if (x<20)
        printf("triple %d\n",p[x]/3);
    else if (x<40)
        printf("double %d\n",p[x]/2);
    else if (x<60)
        printf("single %d\n",p[x]);
    
    return;
}

File: calculatingdartscores.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.TreeMap;

public class calculatingdartscores
{
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int score = sc.nextInt();
        TreeMap<Integer,Node> tm = new TreeMap<Integer,Node>();
        EasyQueue<Node> eq = new EasyQueue<Node>();
        Node root = new Node(score, new int[]{0,0}, 0);
        tm.put(new Integer(root.getValue()), root);
        eq.enqueue(root);
        ArrayList<Node> solution = new ArrayList<Node>();
        DANCE:
        while (true) {
            try {
                Node n = eq.dequeue();
                int v = n.getValue();
                int attempt = n.getAttempt()+1;
                for (int i = 1; i <= 20; i++) {
                    for (int j = 1; j <= 3; j++) {
                        int nuv = v-(i*j);
                        if (nuv == 0) {
                            n = new Node(nuv, new int[]{i,j}, attempt);
                            while (attempt > 0) {
                                solution.add(n);
                                int par = n.getValue()+(n.getPath()[0]*n.getPath()[1]);
                                n = tm.get(new Integer(par));
                                attempt = n.getAttempt();
                            }
                            break DANCE;
                        }
                        else if ((nuv > 0) && (attempt < 3)) {
                            if (tm.get(new Integer(nuv)) == null) {
                                Node nun = new Node(nuv, new int[]{i,j}, attempt);
                                tm.put(new Integer(nuv), nun);
                                eq.enqueue(nun);
                            }
                        }
                    }
                }
            }
            catch (EmptyQueueException e) {
                break DANCE;
            }
        }
        if (solution.size() == 0)
            System.out.println("impossible");
        else {
            for (int i = 0; i < solution.size(); i++) {
                Node n = solution.get(i);
                int[] p = n.getPath();
                switch (p[1]) {
                  case 1:
                    System.out.print("single ");
                    break;
                  case 2:
                    System.out.print("double ");
                    break;
                  case 3:
                    System.out.print("triple ");
                    break;
                }
                System.out.println(p[0]);
            }
        }
    }


    private static class Node {

        private int value;
        private int[] path;
        private int attempt;

        public Node(int v, int[] p, int a) {
            value = v;
            path = new int[2];
            System.arraycopy(p, 0, path, 0, 2);
            attempt = a;
        }

        public int getValue() {
            return value;
        }
        public int[] getPath() {
            return path;
        }
        public int getAttempt() {
            return attempt;
        }

    }

}
